drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,779 @@
+		if (res != 0) {
+			LOG_ERR("Failed to set clock to %d", ios->clock);
+			res = -EIO;
+			goto out;
+		} else {
+			host_io->clock = ios->clock;
+			LOG_DBG("Clock set to %d", ios->clock);
+		}
+	}
+
+	if (ios->power_mode == SDHC_POWER_OFF) {
+		(void)SDMMC_PowerState_OFF(config->hsd->Instance);
+		k_msleep(data->props.power_delay);
+	} else {
+		(void)SDMMC_PowerState_ON(config->hsd->Instance);
+		k_msleep(data->props.power_delay);
+	}
+
+	if ((ios->bus_width != 0) && (host_io->bus_width != ios->bus_width)) {
+		uint32_t bus_width_reg_value = SDMMC_BUS_WIDE_1B;
+
+		if (ios->bus_width == SDHC_BUS_WIDTH8BIT) {
+			bus_width_reg_value = SDMMC_BUS_WIDE_8B;
+		}
+		if (ios->bus_width == SDHC_BUS_WIDTH4BIT) {
+			bus_width_reg_value = SDMMC_BUS_WIDE_4B;
+		}
+
+		MODIFY_REG(config->hsd->Instance->CLKCR, SDMMC_CLKCR_WIDBUS, bus_width_reg_value);
+		host_io->bus_width = ios->bus_width;
+	}
+
+out:
+	k_mutex_unlock(&data->bus_mutex);
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	return res;
+}
+
+static void sdhc_stm32_init_props(const struct device *dev)
+{
+	const struct sdhc_stm32_config *sdhc_config = (const struct sdhc_stm32_config *)dev->config;
+	struct sdhc_stm32_data *data = dev->data;
+	struct sdhc_host_props *props = &data->props;
+
+	memset(props, 0, sizeof(struct sdhc_host_props));
+
+	props->f_min = sdhc_config->min_freq;
+	props->f_max = sdhc_config->max_freq;
+	props->power_delay = sdhc_config->power_delay_ms;
+	props->host_caps.vol_330_support = true;
+	props->host_caps.vol_180_support = sdhc_config->support_1_8_v;
+	props->host_caps.bus_8_bit_support = (sdhc_config->bus_width == SDHC_BUS_WIDTH8BIT);
+	props->host_caps.bus_4_bit_support = (sdhc_config->bus_width == SDHC_BUS_WIDTH4BIT);
+}
+
+static int sdhc_stm32_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	struct sdhc_stm32_data *data = dev->data;
+
+	memcpy(props, &data->props, sizeof(struct sdhc_host_props));
+
+	return 0;
+}
+
+static int sdhc_stm32_get_card_present(const struct device *dev)
+{
+	int res;
+	struct sdhc_stm32_data *dev_data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&dev_data->bus_mutex, K_FOREVER);
+
+	/* If a CD pin is configured, use it for card detection */
+	if (config->cd_gpio.port != NULL) {
+		if (!device_is_ready(config->cd_gpio.port)) {
+			LOG_ERR("Card detect GPIO device not ready");
+			res = -ENODEV;
+			goto out;
+		}
+		res = gpio_pin_get_dt(&config->cd_gpio);
+	} else {
+		/* Card is considered present if the read command did not time out */
+		res = SDMMC_CmdSendOperationcondition(config->hsd->Instance, 0, NULL);
+		if (res != 0) {
+			sdhc_stm32_log_err_type(config->hsd);
+		}
```

Comment:
`res` is of typee `int`.
```suggestion
		if (SDMMC_CmdSendOperationcondition(config->hsd->Instance, 0, NULL) != 0) {
			res = -EIO;
			sdhc_stm32_log_err_type(config->hsd);
		}